File: hdar/cpp/src/controller.cpp
#include "hdar/controller.hpp"

namespace hdar {

LifecycleController::LifecycleController(
    const AgentIdentity& identity,
    ProviderBase* provider,
    ContentStore& store,
    const ControllerConfig& config)
    : identity_(identity)
    , provider_(provider)
    , store_(store)
    , config_(config)
    , sm_(identity.agent_id)
    , effects_(config.effects_ledger, nullptr, identity.agent_id)
    , lease_manager_(config.lease_db, config.lease_ttl)
    , sealer_(store, const_cast<AgentIdentity&>(identity), &lease_manager_) {}

std::pair<RuntimeRecord, std::optional<std::string>> LifecycleController::wake(
    const std::string& capsule_hash,
    const std::string& workspace_dir,
    const std::string& destination_runtime) {

    sm_.transition(AgentState::ACQUIRING_LEASE, "wake requested");

    std::string runtime_name = destination_runtime.empty()
        ? ("runtime-" + generate_uuid_hex().substr(0, 8))
        : destination_runtime;

    auto [lease, error] = lease_manager_.acquire(
        identity_.agent_id, capsule_hash, 0,
        identity_.agent_id, runtime_name);

    if (error) {
        sm_.transition(AgentState::LEASE_LOST, *error);
        return {{}, *error};
    }

    lease_ = lease;
    fencing_token_ = lease->fencing_token;

    // Re-create effects registry with lease manager now that we have a lease
    effects_ = EffectRegistry(config_.effects_ledger, &lease_manager_, identity_.agent_id);

    sm_.transition(AgentState::MATERIALIZING, "lease acquired gen=" +
                   std::to_string(lease->lease_generation));

    current_runtime_id_ = runtime_name;
    auto runtime = provider_->materialize(runtime_name, workspace_dir);
    runtime_ = runtime;

    if (!runtime.exists) {
        sm_.transition(AgentState::RESTORE_REJECTED, "materialization failed");
        return {runtime, "materialization failed"};
    }

    sm_.transition(AgentState::VERIFYING_INPUT, "runtime materialized");
    sm_.transition(AgentState::RUNNING, "verification passed");

    return {runtime, std::nullopt};
}

ExecutionResult LifecycleController::execute(
    const std::string& operation_type,
    const std::string& command,
    int timeout) {

    if (!sm_.is_running())
        throw std::runtime_error("cannot execute: agent not in RUNNING state");

    return provider_->execute(current_runtime_id_, operation_type, command, timeout);
}

ExternalEffect LifecycleController::register_effect(
    const std::string& capability,
    const std::vector<uint8_t>& payload,
    const std::string& operation_id) {

    return effects_.register_effect(
        identity_.agent_id, capability, payload, operation_id, fencing_token_);
}

ExternalEffect LifecycleController::commit_effect(
    const std::string& operation_id,
    const JsonValue& provider_receipt) {

    return effects_.commit(
        identity_.agent_id, operation_id, provider_receipt, fencing_token_);
}

ExternalEffect LifecycleController::mark_effect_unknown(const std::string& operation_id) {
    return effects_.mark_unknown(identity_.agent_id, operation_id);
}

JsonValue LifecycleController::check_quiescence() const {
    return effects_.check_quiescence(identity_.agent_id);
}

JsonValue LifecycleController::collapse(
    const LineageEpoch& epoch,
    const std::string& objective,
    const std::string& continuation_point,
    const std::string& working_summary,
    const JsonValue& capabilities,
    const std::string& capability_note,
    const std::optional<std::string>& parent_capsule_hash,
    const std::string& fencing_token) {

    // 0. Validate fencing token — stale holder cannot collapse
    if (!fencing_token.empty()) {
        if (!lease_manager_.validate_token(identity_.agent_id, fencing_token)) {
            sm_.transition(AgentState::LEASE_LOST,
                           "stale fencing token — cannot collapse");
            JsonValue result = JsonValue::object();
            result["collapsed"] = JsonValue::boolean(false);
            result["reason"] = JsonValue::string(
                "stale or invalid fencing token — this runtime's lease "
                "is no longer authoritative");
            result["runtime_id"] = JsonValue::string(current_runtime_id_);
            return result;
        }
    }

    // 1. Check quiescence
    auto q = check_quiescence();
    if (!q.get("quiescent").bool_val) {
        sm_.transition(AgentState::UNKNOWN_EFFECT, "blocking effects prevent seal");
        JsonValue result = JsonValue::object();
        result["collapsed"] = JsonValue::boolean(false);
        result["reason"] = JsonValue::string("blocking effects prevent seal");
        result["quiescence"] = q;
        return result;
    }

    // 2. Transition to quiescing
    sm_.transition(AgentState::QUIESCING, "quiescent");

    // 3. Seal capsule
    sm_.transition(AgentState::SEALING, "sealing capsule");

    std::string workspace_dir = runtime_ && !runtime_->workspace_mount.empty()
        ? runtime_->workspace_mount
        : config_.workspace_root + "/" + current_runtime_id_;
    auto [manifest, chain] = sealer_.seal(
        workspace_dir, epoch, objective, continuation_point, working_summary,
        capabilities, capability_note, parent_capsule_hash,
        "", "", JsonValue::object(), JsonValue::array(), JsonValue::array(),
        JsonValue::object(), fencing_token_);

    // 4. Write capsule
    std::string capsule_path = config_.store_dir + "/capsules/" +
                               manifest.manifest_hash.substr(0, 16) + ".json";
    sealer_.write_capsule(manifest, capsule_path);

    // 5. Destroy runtime
    sm_.transition(AgentState::DESTROYING, "destroying runtime");
    auto destroyed = provider_->destroy(current_runtime_id_);

    // 6. Verify destruction
    bool destruction_confirmed = provider_->verify_destruction(current_runtime_id_);
    if (!destruction_confirmed) {
        sm_.transition(AgentState::DESTRUCTION_UNCONFIRMED, "destruction not confirmed");
    } else {
        sm_.transition(AgentState::DORMANT, "destruction confirmed, dormant");
    }

    // 7. Release lease
    lease_manager_.release(identity_.agent_id, fencing_token_);
    lease_ = std::nullopt;
    fencing_token_.clear();
    runtime_ = std::nullopt;

    JsonValue result = JsonValue::object();
    result["collapsed"] = JsonValue::boolean(true);
    result["manifest_hash"] = JsonValue::string(manifest.manifest_hash);
    result["capsule_path"] = JsonValue::string(capsule_path);
    result["destruction_confirmed"] = JsonValue::boolean(destruction_confirmed);
    result["final_state"] = JsonValue::string(state_name(sm_.state()));
    result["manifest"] = manifest.to_json();
    return result;
}

} // namespace hdar

File: hdar/lifecycle/controller.py
"""Lifecycle controller — orchestrates the full agent lifecycle.

Ties together:
  - State machine (P0 #1)
  - Effect registry (P0 #2)
  - Fenced lease manager (P0 #3)
  - Execution provider (P0 #4)

Enforces the rule: a capsule may seal only after quiescence and
effect reconciliation. Runtime destruction must be verified before
returning to DORMANT.
"""

from __future__ import annotations

import json
import os
import time
import uuid
from pathlib import Path
from typing import Optional

from .state_machine import LifecycleStateMachine, AgentState
from .effects import EffectRegistry
from .lease import LeaseManager
from providers.base import ProviderBase, RuntimeRecord, ExecutionResult


class LifecycleController:
    """Orchestrates: wake → run → quiesce → seal → destroy → dormant."""

    def __init__(
        self,
        agent_id: str,
        provider: ProviderBase,
        lease_manager: LeaseManager,
        effect_registry: EffectRegistry,
        state_dir: str,
    ):
        self.agent_id = agent_id
        self.provider = provider
        self.lease_manager = lease_manager
        self.effects = effect_registry
        self.sm = LifecycleStateMachine(agent_id)
        self.state_dir = Path(state_dir)
        self.state_dir.mkdir(parents=True, exist_ok=True)

        self.lease = None
        self.runtime_id: Optional[str] = None
        self.runtime_record: Optional[RuntimeRecord] = None

    def wake(
        self,
        capsule_hash: str,
        epoch: int,
        workspace_path: str,
        holder_id: str = "local",
        image: str = "",
        cpu_limit: str = "2",
        memory_limit: str = "2g",
    ) -> dict:
        """Wake the agent: acquire lease, materialize runtime, verify."""
        if not self.sm.transition(AgentState.ACQUIRING_LEASE, "wake requested"):
            return {"woke": False, "reason": f"cannot wake from {self.sm.state.name}"}

        # Acquire exclusive lease
        self.lease, err = self.lease_manager.acquire(
            self.agent_id, capsule_hash, epoch, holder_id, "pending"
        )
        if err:
            self.sm.transition(AgentState.LEASE_LOST, err)
            return {"woke": False, "reason": f"lease denied: {err}"}

        # Materialize runtime
        if not self.sm.transition(AgentState.MATERIALIZING, "provider selected"):
            return {"woke": False, "reason": "state machine rejected materialize"}

        self.runtime_id = f"rt-{uuid.uuid4().hex[:8]}"
        self.runtime_record = self.provider.materialize(
            runtime_id=self.runtime_id,
            workspace_path=workspace_path,
            image=image,
            cpu_limit=cpu_limit,
            memory_limit=memory_limit,
        )

        # Verify input
        if not self.sm.transition(AgentState.VERIFYING_INPUT, "runtime materialized"):
            return {"woke": False, "reason": "state machine rejected verify"}

        # Check runtime exists
        inspection = self.provider.inspect(self.runtime_id)
        if not inspection.get("exists"):
            self.sm.transition(AgentState.RESTORE_REJECTED, "runtime not found after materialize")
            return {"woke": False, "reason": "runtime not found after materialize"}

        # Running
        self.sm.transition(AgentState.RUNNING, "capsule verified, runtime active")
        return {
            "woke": True,
            "agent_id": self.agent_id,
            "runtime_id": self.runtime_id,
            "lease_generation": self.lease.lease_generation,
            "fencing_token": self.lease.fencing_token,
            "provider": self.provider.name,
            "state": self.sm.state.name,
        }

    def execute(self, operation_type: str, command: str, timeout: int = 60) -> ExecutionResult:
        """Execute a typed operation inside the runtime."""
        if not self.sm.is_running():
            return ExecutionResult(
                operation_type=operation_type, command=command,
                exit_code=-1, stderr="agent not running", success=False,
            )
        return self.provider.execute(self.runtime_id, operation_type, command, timeout)

    def register_effect(self, capability: str, payload: bytes,
                        operation_id: Optional[str] = None):
        """Register an external effect before executing it."""
        return self.effects.register(
            self.agent_id, capability, payload, operation_id=operation_id
        )

    def commit_effect(self, operation_id: str, provider_receipt: Optional[dict] = None):
        """Mark an external effect as committed."""
        return self.effects.commit(self.agent_id, operation_id, provider_receipt=provider_receipt)

    def mark_effect_unknown(self, operation_id: str):
        """Mark an external effect as unknown (e.g. crash before confirmation)."""
        return self.effects.mark_unknown(self.agent_id, operation_id)

    def reconcile_effects(self, probe_fn) -> dict:
        """Reconcile unknown effects on wake."""
        return self.effects.reconcile(self.agent_id, probe_fn)

    def collapse(self, force: bool = False, fencing_token: str = "") -> dict:
        """Collapse: quiesce → seal → destroy → dormant.

        GATED on semantic quiescence. Refuses to seal while external
        effects are in flight, unless force=True (for emergencies).

        When a fencing token is provided, it is validated against the
        current lease. A stale runtime cannot collapse or release the lease.
        """
        if fencing_token and self.lease:
            if not self.lease_manager.validate_token(self.agent_id, fencing_token):
                self.sm.transition(AgentState.LEASE_LOST,
                                   "stale fencing token — cannot collapse")
                return {
                    "collapsed": False,
                    "reason": "stale or invalid fencing token — this runtime's lease is no longer authoritative",
                    "runtime_id": self.runtime_id,
                }
        # Allow collapse from RUNNING or UNKNOWN_EFFECT (after reconciliation)
        if self.sm.state not in (AgentState.RUNNING, AgentState.UNKNOWN_EFFECT):
            return {"collapsed": False, "reason": f"agent is {self.sm.state.name}, not running or quiescible"}

        # If in UNKNOWN_EFFECT, transition back to QUIESCING after reconciliation
        if self.sm.state == AgentState.UNKNOWN_EFFECT:
            if not self.sm.transition(AgentState.QUIESCING, "effects reconciled, retrying collapse"):
                return {"collapsed": False, "reason": "cannot transition from UNKNOWN_EFFECT to QUIESCING"}
        else:
            # Quiesce from RUNNING
            self.sm.transition(AgentState.QUIESCING, "collapse requested")

        # Check quiescence
        q = self.effects.check_quiescence(self.agent_id)
        if not q["quiescent"] and not force:
            self.sm.transition(AgentState.UNKNOWN_EFFECT, q["verdict"])
            return {
                "collapsed": False,
                "reason": q["verdict"],
                "blocking_effects": q["blocking_effects"],
            }

        # Seal
        self.sm.transition(AgentState.SEALING, "quiescence confirmed")

        # Destroy runtime
        self.sm.transition(AgentState.DESTROYING, "sealing complete")
        if self.runtime_id:
            self.provider.stop(self.runtime_id)
            destroy_record = self.provider.destroy(self.runtime_id)

            # Verify destruction
            destroyed = self.provider.verify_destruction(self.runtime_id)
            if not destroyed:
                self.sm.transition(AgentState.DESTRUCTION_UNCONFIRMED,
                                   "runtime still exists after destroy")
                return {
                    "collapsed": False,
                    "reason": "DESTRUCTION UNCONFIRMED — runtime still exists",
                    "runtime_id": self.runtime_id,
                }

        # Release lease
        if self.lease:
            self.lease_manager.release(self.agent_id, self.lease.fencing_token)
            self.lease = None

        self.runtime_id = None
        self.runtime_record = None
        self.sm.transition(AgentState.DORMANT, "runtime destroyed, lease released")

        return {
            "collapsed": True,
            "agent_id": self.agent_id,
            "runtime_destroyed": True,
            "destruction_verified": True,
            "active_compute": "zero — dormant storage only",
            "state": self.sm.state.name,
        }

    def status(self) -> dict:
        return {
            "agent_id": self.agent_id,
            "state": self.sm.state.name,
            "is_running": self.sm.is_running(),
            "is_dormant": self.sm.is_dormant(),
            "runtime_id": self.runtime_id,
            "provider": self.provider.name,
            "lease_generation": self.lease.lease_generation if self.lease else None,
            "transitions": len(self.sm.history),
        }
